README.md
# aws_public_ip_addresses
Amazon AWS publishes the list of public IP block they own organized by their various regions.  For thos of us who want to use that list, it's a bit big to handle manually.  This project is a simple tool for working with it.


First, the resources:
Amazon desribes the list here:
https://docs.aws.amazon.com/general/latest/gr/aws-ip-ranges.html


That page has a link to download the JSON file:

https://ip-ranges.amazonaws.com/ip-ranges.json



CHANGELOG.rst

Changelog
=========

0.0.0 (2019-01-24)
------------------

* First release on PyPI.


tests/test_model.py
from aws_public_ip_addresses.model import AWSAddresses, Renderer, CiscoASA


def test_check_download():
    """
    We expect the 'prefixes' key to exist at the top of the json download object
    """
    assert 'prefixes' in AWSAddresses.download_and_get_data()


def test_get_data():
    addresses = AWSAddresses()
    output = addresses.get_data()
    assert 'prefixes' in output


def test_get_ipv4_prefixes():
    addresses = AWSAddresses()
    assert len(addresses.get_ipv4_prefixes()) > 0


def test_get_ipv6_prefixes():
    addresses = AWSAddresses()
    assert len(addresses.get_ipv6_prefixes()) > 0


def test_get_regions():
    """
    :return:
    """
    addresses = AWSAddresses()

    all_regions = ['GLOBAL', 'ap-northeast-1', 'ap-northeast-2', 'ap-northeast-3', 'ap-south-1', 'ap-southeast-1',
                   'ap-southeast-2', 'ca-central-1', 'cn-north-1', 'cn-northwest-1', 'eu-central-1', 'eu-north-1',
                   'eu-west-1', 'eu-west-2', 'eu-west-3', 'me-south-1', 'sa-east-1', 'us-east-1', 'us-east-2',
                   'us-gov-east-1', 'us-gov-west-1', 'us-west-1', 'us-west-2']

    all_regions.sort()

    ipv4_regions = ['GLOBAL', 'ap-northeast-1', 'ap-northeast-2', 'ap-northeast-3', 'ap-south-1', 'ap-southeast-1',
                    'ap-southeast-2', 'ca-central-1', 'cn-north-1', 'cn-northwest-1', 'eu-central-1', 'eu-north-1',
                    'eu-west-1', 'eu-west-2', 'eu-west-3', 'me-south-1', 'sa-east-1', 'us-east-1', 'us-east-2',
                    'us-gov-east-1', 'us-gov-west-1', 'us-west-1', 'us-west-2']

    ipv4_regions.sort()

    ipv6_regions = ['GLOBAL', 'ap-northeast-1', 'ap-northeast-2', 'ap-northeast-3', 'ap-south-1', 'ap-southeast-1',
                    'ap-southeast-2', 'ca-central-1', 'cn-north-1', 'cn-northwest-1', 'eu-central-1', 'eu-north-1',
                    'eu-west-1', 'eu-west-2', 'eu-west-3', 'me-south-1', 'sa-east-1', 'us-east-1', 'us-east-2',
                    'us-gov-east-1', 'us-gov-west-1', 'us-west-1', 'us-west-2']

    ipv6_regions.sort()

    assert all_regions == addresses.get_regions()
    assert ipv4_regions == addresses.get_regions(ip_version=4)
    assert ipv6_regions == addresses.get_regions(ip_version=6)


def test_filter_ipv4_prefixes():
    addresses = AWSAddresses()
    assert len(addresses.filter_ipv4_prefixes('us-east-')) > 0


def test_filter_ipv6_prefixes():
    addresses = AWSAddresses()
    assert len(addresses.filter_ipv6_prefixes('us-east-')) > 0


def test_renderer_class_factory():
    addresses = AWSAddresses()
    template_data = {'ipv4_addresses': addresses.filter_ipv4_prefixes('us-east-'),
                     'ipv6_addresses': addresses.filter_ipv6_prefixes('us-east-')}
    assert isinstance(Renderer.get_renderer(template_data, 'CiscoASA'), CiscoASA)


def test_cisco_object_data():
    assert '192_168_1_0_S_24', 'subnet 192.168.1.0 255.255.255.0' == \
           CiscoASA.cisco_object_data('192.168.1.0/24')


tests/test_aws_public_ip_addresses.py

from click.testing import CliRunner

from aws_public_ip_addresses.cli import main


def test_main():
    runner = CliRunner()
    result = runner.invoke(main, [])

    assert result.output == '()\n'
    assert result.exit_code == 0


ci/appveyor-download.py
#!/usr/bin/env python
"""
Use the AppVeyor API to download Windows artifacts.

Taken from: https://bitbucket.org/ned/coveragepy/src/tip/ci/download_appveyor.py
# Licensed under the Apache License: http://www.apache.org/licenses/LICENSE-2.0
# For details: https://bitbucket.org/ned/coveragepy/src/default/NOTICE.txt
"""
from __future__ import unicode_literals

import argparse
import os
import zipfile

import requests


def make_auth_headers():
    """Make the authentication headers needed to use the Appveyor API."""
    path = os.path.expanduser("~/.appveyor.token")
    if not os.path.exists(path):
        raise RuntimeError(
            "Please create a file named `.appveyor.token` in your home directory. "
            "You can get the token from https://ci.appveyor.com/api-token"
        )
    with open(path) as f:
        token = f.read().strip()

    headers = {
        'Authorization': 'Bearer {}'.format(token),
    }
    return headers


def download_latest_artifacts(account_project, build_id):
    """Download all the artifacts from the latest build."""
    if build_id is None:
        url = "https://ci.appveyor.com/api/projects/{}".format(account_project)
    else:
        url = "https://ci.appveyor.com/api/projects/{}/build/{}".format(account_project, build_id)
    build = requests.get(url, headers=make_auth_headers()).json()
    jobs = build['build']['jobs']
    print(u"Build {0[build][version]}, {1} jobs: {0[build][message]}".format(build, len(jobs)))

    for job in jobs:
        name = job['name']
        print(u"  {0}: {1[status]}, {1[artifactsCount]} artifacts".format(name, job))

        url = "https://ci.appveyor.com/api/buildjobs/{}/artifacts".format(job['jobId'])
        response = requests.get(url, headers=make_auth_headers())
        artifacts = response.json()

        for artifact in artifacts:
            is_zip = artifact['type'] == "Zip"
            filename = artifact['fileName']
            print(u"    {0}, {1} bytes".format(filename, artifact['size']))

            url = "https://ci.appveyor.com/api/buildjobs/{}/artifacts/{}".format(job['jobId'], filename)
            download_url(url, filename, make_auth_headers())

            if is_zip:
                unpack_zipfile(filename)
                os.remove(filename)


def ensure_dirs(filename):
    """Make sure the directories exist for `filename`."""
    dirname = os.path.dirname(filename)
    if dirname and not os.path.exists(dirname):
        os.makedirs(dirname)


def download_url(url, filename, headers):
    """Download a file from `url` to `filename`."""
    ensure_dirs(filename)
    response = requests.get(url, headers=headers, stream=True)
    if response.status_code == 200:
        with open(filename, 'wb') as f:
            for chunk in response.iter_content(16 * 1024):
                f.write(chunk)
    else:
        print(u"    Error downloading {}: {}".format(url, response))


def unpack_zipfile(filename):
    """Unpack a zipfile, using the names in the zip."""
    with open(filename, 'rb') as fzip:
        z = zipfile.ZipFile(fzip)
        for name in z.namelist():
            print(u"      extracting {}".format(name))
            ensure_dirs(name)
            z.extract(name)


parser = argparse.ArgumentParser(description='Download artifacts from AppVeyor.')
parser.add_argument('--id',
                    metavar='PROJECT_ID',
                    default='natemarks/aws_public_ip_addresses',
                    help='Project ID in AppVeyor.')
parser.add_argument('build',
                    nargs='?',
                    metavar='BUILD_ID',
                    help='Build ID in AppVeyor. Eg: master-123')

if __name__ == "__main__":
    # import logging
    # logging.basicConfig(level="DEBUG")
    args = parser.parse_args()
    download_latest_artifacts(args.id, args.build)


src/aws_public_ip_addresses/model.py
from typing import List, Dict, Any, Tuple


class AWSAddresses(object):
    """AWS Public IP Addresses

    If the class has public attributes, they may be documented here
    in an ``Attributes`` section and follow the same formatting as a
    function's ``Args`` section. Alternatively, attributes may be documented
    inline with the attribute's declaration (see __init__ method below).

    """
    json_download_url = 'https://ip-ranges.amazonaws.com/ip-ranges.json'  # type: str
    ipv4_prefix_key = 'prefixes'  # type: str
    ipv6_prefix_key = 'ipv6_prefixes'  # type: str

    def __init__(self, *args, **kwargs):
        self.args = args
        self.kwargs = {}
        self.kwargs.update(kwargs)
        self._regions = None  # type: List[str]
        self._download_data = None
        self._ipv4_prefixes = None  # type: List[Dict[str, str]]
        self._ipv6_prefixes = None  # type: List[Dict[str, str]]

    @classmethod
    def download_and_get_data(cls) -> Dict[str, Any]:
        """Force download and return the json as nested dict

        """
        from urllib.request import urlopen
        import json

        json_data = urlopen(AWSAddresses.json_download_url).read()

        return json.loads(json_data)

    def get_data(self):
        if not self._download_data:
            self._download_data = AWSAddresses.download_and_get_data()
        return self._download_data

    def get_ipv4_prefixes(self) -> List[Dict[str, str]]:
        """The summary line for a method docstring should fit on one line.

        returns the ipv4 prefixes dict
        """
        if not self._ipv4_prefixes:
            self._ipv4_prefixes = self.get_data()[AWSAddresses.ipv4_prefix_key]
        return self._ipv4_prefixes

    def get_ipv6_prefixes(self) -> List[Dict[str, str]]:
        """The summary line for a method docstring should fit on one line.

        returns the ipv6 prefixes dict
        """
        if not self._ipv6_prefixes:
            self._ipv6_prefixes = self.get_data()[AWSAddresses.ipv6_prefix_key]
        return self._ipv6_prefixes

    def get_regions(self, ip_version: int = None) -> List[str]:
        """Return a list of the regions

        if ip_version is left default (None), the method will return a list of all unique reqions from both the ipv4
        and ipv6 prefix dicts,  If ip_version=4 is specified, it will just use the ipv4 prefix dict and if ip_version=6
        is specified it'll just use the ipv6 dict

        """
        if not self._regions:
            region_list = []
            if ip_version in [4, None]:
                for prefix in self.get_ipv4_prefixes():
                    if prefix['region'] not in region_list:
                        region_list.append(prefix['region'])

            if ip_version in [6, None]:
                for prefix in self.get_ipv6_prefixes():
                    if prefix['region'] not in region_list:
                        region_list.append(prefix['region'])

            region_list.sort()
            self._regions = region_list

        return self._regions

    def filter_ipv4_prefixes(self, regions_starting_with: str) -> List[str]:
        """Return filtered list of ipv4 prefixes

        Require regions_starting_with  and just return the prefixes for regions that start with that string.
        ex.  if regions_starting_with == 'us-east-'
        we'll return the us-east-1  and us-east-2 prefixes, but not us-west-2


        """
        ipv4_prefixes = self.get_ipv4_prefixes()
        results = []
        for prefix in ipv4_prefixes:
            if prefix['region'].startswith(regions_starting_with):
                results.append(prefix['ip_prefix'])

        return results

    def filter_ipv6_prefixes(self, regions_starting_with: str) -> List[str]:
        """Return filtered list of ipv6 prefixes

        Require regions_starting_with  and just return the prefixes for regions that start with that string.
        ex.  if regions_starting_with == 'us-east-'
        we'll return the us-east-1  and us-east-2 prefixes, but not us-west-2


        """
        ipv6_prefixes = self.get_ipv6_prefixes()
        results = []
        for prefix in ipv6_prefixes:
            if prefix['region'].startswith(regions_starting_with):
                results.append(prefix['ipv6_prefix'])

        return results


class Renderer(object):
    """render the AWS addresses in various formats

    If the class has public attributes, they may be documented here
    in an ``Attributes`` section and follow the same formatting as a
    function's ``Args`` section. Alternatively, attributes may be documented
    inline with the attribute's declaration (see __init__ method below).

    """
    def __init__(self, addresses: Dict[str, List[str]]):
        """
        
        :type addresses: Dict[str, List[str]]
        :param addresses: This should be dict with two keys:  one for ipv4 addresses and one for ipv6 addresses

        {'ipv4_addresses': ['192.168.1.0/24', '10.0.0.0/16'],
        'ipv6_addresses': ['....', '....']}
        """
        self.addresses = addresses

    @classmethod
    def get_renderer(cls, aws_addresses: AWSAddresses, render_format='CiscoASA'):
        if render_format == 'CiscoASA':
            return CiscoASA(aws_addresses)
        raise RuntimeError('Invalid Rendere Class: ' + render_format)

    @staticmethod
    def render_template(jinja_template: str = None, template_data=None) -> str:
        """Given a dict, render the template and return as string

        Note that PackageLoader lets me put the template in src and from the template in the package after install
        """
        from jinja2 import Environment, PackageLoader
        env = Environment(loader=PackageLoader('aws_public_ip_addresses', 'templates'),
                          trim_blocks=True, lstrip_blocks=True)
        template = env.get_template(jinja_template)

        return template.render(template_data)


class CiscoASA(Renderer):
    """Renders group objects from prefix list for Cisco ASA

    If the class has public attributes, they may be documented here
    in an ``Attributes`` section and follow the same formatting as a
    function's ``Args`` section. Alternatively, attributes may be documented
    inline with the attribute's declaration (see __init__ method below).

    """

    def __init__(self, addresses: Dict[str, List[str]]):
        super().__init__(addresses)

        self.ipv4_object_group_name = 'AMAZON_AWS_IPV4_ADDRESSES'
        self.ipv6_object_group_name = 'AMAZON_AWS_IPV6_ADDRESSES'

        template_file = 'cisco_asa.jinja2'
        template_data = {
            'ipv4_group_name': self.ipv4_object_group_name,
            'ipv6_group_name': self.ipv6_object_group_name,
            'ipv4_addresses': CiscoASA.get_object_dict(addresses['ipv4_addresses']),
            'ipv6_addresses': CiscoASA.get_object_dict(addresses['ipv6_addresses'])
        }

        self.output = Renderer.render_template(jinja_template=template_file, template_data=template_data)

    @staticmethod
    def cisco_object_data(ip_prefix: str) -> Tuple[str, str]:
        """Get S Notation and subnet format from slash format prefix
        '192.168.1.0/24' -> '192_168_1_0_S_24', 'subnet 192.168.1.0 255.255.255.0'

        """
        from netaddr import IPNetwork
        net = IPNetwork(ip_prefix)
        obj_data = ' '.join(['subnet', str(net.network), str(net.netmask)])

        for r in ((".", "_"), ("/", "_S_")):
            s_notation = ip_prefix.replace(*r)

        return s_notation, obj_data

    @staticmethod
    def get_object_dict(prefix_list: List[str]) -> Dict[str, str]:
        result_dict = {}
        for prefix in prefix_list:
            k, v = CiscoASA.cisco_object_data(prefix)
            result_dict[k] = v
        return result_dict


